libs/zauth/src/Data/ZAuth/Token.hs

```diff
@@ -163,33 +204,41 @@ mkBot = Bot
 mkProvider :: UUID -> Provider
 mkProvider = Provider
 
+mkLegalHoldAccess :: UUID -> Word64 -> LegalHoldAccess
+mkLegalHoldAccess uid cid = LegalHoldAccess $ Access uid cid
+
+mkLegalHoldUser :: UUID -> Word32 -> LegalHoldUser
+mkLegalHoldUser uid r = LegalHoldUser $ User uid r
+
 -----------------------------------------------------------------------------
 -- Reading
 
-readToken :: (Properties -> Maybe a) -> LByteString -> Maybe (Token a)
-readToken f b = case split '.' b of
+readToken :: Type -> (Properties -> Maybe a) -> LByteString -> Maybe (Token a)
+readToken t f b = case split '.' b of
     (s:rest) ->
         let p = map pairwise rest in
         Token <$> hush (Signature <$> decode (toStrict s))
-              <*> readHeader p
+              <*> readHeader t p
               <*> f p
     _ -> Nothing
   where
     pairwise :: LByteString -> (LByteString, LByteString)
     pairwise x = let (k, v) = break (== '=') x in (k, drop 1 v)
 
-readHeader :: Properties -> Maybe Header
-readHeader p = Header
+readHeader :: Type -> Properties -> Maybe Header
+readHeader t p = Header
     <$> (lookup "v" p >>= fromByteString')
     <*> (lookup "k" p >>= fromByteString')
     <*> (lookup "d" p >>= fromByteString')
-    <*> (lookup "t" p >>= readType)
+    <*> (lookup "t" p >>= readType >>= \t' -> if t == t' then Just t' else Nothing)
```

Comment:
nit-pick:

```haskell
    <*> (lookup "t" p >>= readType t)
[...]
    readType A "a" = Just A
    readType U "u" = Just U
    readType B "b" = Just B
    readType P "p" = Just P
    readType LA "la" = Just LA
    readType LU "lu" = Just LU
    readType _  _ = Nothing
```

Reply:
I also think that `Header` should be this:

```haskell
data Header (t :: Type) = Header
    { _version   :: !Int
    , _key       :: !Int
    , _time      :: !Integer
    , _tag       :: Maybe Tag
    } deriving (Eq, Show)
```

This is also not part of the feature, but I'm tempted to request that this is included anyway.  (Out of curiosity: did `readHeader` start taking a type argument because it didn't work any more without?  Or is that just an extra check that we *can* do, but neglected to do so far?)

Reply:
> is that just an extra check that we can do, but neglected to do so far?

Exactly, the type header was never checked, leading to issues, since LegalHold and regular tokens are the same *except* for their type header.

Reply:
> nit-pick:
> 
> ```haskell
>     <*> (lookup "t" p >>= readType t)
> [...]
>     readType A "a" = Just A
>     readType U "u" = Just U
>     readType B "b" = Just B
>     readType P "p" = Just P
>     readType LA "la" = Just LA
>     readType LU "lu" = Just LU
>     readType _  _ = Nothing
> ```

Thanks, that's clearer!